libs/types/src/auth/dipped_smart_card_auth.ts

```diff
@@ -21,7 +21,10 @@ export interface CheckingPin {
   readonly status: 'checking_pin';
   readonly user: SystemAdministratorUser | ElectionManagerUser;
   readonly error?: true;
-  readonly wrongPinEnteredAt?: Date;
+  /** A Unix timestamp in milliseconds for easy serialization */
+  readonly lockedOutUntil?: number;
```

Comment:
I encountered some interesting serialization bugs when I was using Dates. `JSON.stringify` automatically converts Dates to strings. So when I was using a Date type here, the type was inaccurate on the frontend, which was actually working with a string.

I looked into updating Grout's serialization to handle Dates, but Grout's serialization special-casing works best for types that can't be serialized at all. Dates actually are serialized by `JSON.stringify`, so I'd have to do some guess work to gauge whether a value was originally a Date before it became a string but then run the risk of accidentally converting some values that were meant to be strings into Dates. Ultimately decided that finding a general solution was out of scope, so I switched to Unix timestamps in milliseconds instead. Will open an issue for this! cc @jonahkagan

Reply:
Did a little digging on this because I was surprised. It turns out that the `JSON.stringify` replacer function (which is what Grout uses for tagging values during serialization) is called with a transformed value, not the original value. However, you can still access the original value using `this[key]`. Wild!

```ts
> JSON.stringify(new Date(), function (k, v) { console.log(typeof v, typeof this[k]); return this[k] instanceof Date ? 1 : v})
string object
'1'
```

I think this should allow us to catch and handle dates.

Alternatively we could just write our own recursive tagger and then call `JSON.stringify` on the resulting transformed value. I considered this originally but decided to try out the built-in replacer functionality to see how it worked.

I do think it's worth fixing because this is a potential footgun for anyone who tries to use a Date.

Reply:
https://github.com/votingworks/vxsuite/pull/3239

Reply:
Thanks for digging into this!